File: manipulation/src/manipulation_request.cpp
#include <iostream>
#include "manipulation/manipulation.h"
#include <manipulation/GetManipulationPlan.h>
#include <manipulation/ManipulationPlanRequest.h>
#include <actionlib/client/simple_action_client.h>

constexpr char LOGNAME[] = "manipulation request";

using namespace manipulation;

int main(int argc, char* argv[])
{
  ros::init(argc, argv, "manipulation_request");
  ros::NodeHandle nh, pnh("~");

  ros::AsyncSpinner spinner(1);
  spinner.start();

  ros::Duration(1.0).sleep();

  GetManipulationPlan get_plan;
  get_plan.request.manipulation_plan_request.task_type = ManipulationPlanRequest::DANCE; 

  get_plan.request.manipulation_plan_request.object_name = "demo_cube";
  get_plan.request.manipulation_plan_request.support_surfaces = {"table1"};

  get_plan.request.manipulation_plan_request.task_name = "pick_and_place";

  geometry_msgs::Pose place_pose; 
  place_pose.position.x = .6;
  place_pose.position.y = .11;
  place_pose.position.z = .742311;
  place_pose.orientation.w = 1.0;  
 
  get_plan.request.manipulation_plan_request.place_pose = place_pose; 

  ros::ServiceClient client = nh.serviceClient<manipulation::GetManipulationPlanRequest>("get_manipulation_plan");

  actionlib::SimpleActionClient<moveit_task_constructor_msgs::ExecuteTaskSolutionAction>
  execute("execute_task_solution", true); execute.waitForServer();
  
  moveit_task_constructor_msgs::ExecuteTaskSolutionGoal execute_goal;

  if (client.call(get_plan))
  {
    ROS_INFO("Received");
    execute_goal.solution = get_plan.response.manipulation_plan_response.solution;
    execute.sendGoalAndWait(execute_goal);
    moveit_msgs::MoveItErrorCodes execute_result = execute.getResult()->error_code;
    ROS_INFO("Executing Plan");
  }
  else
  {
    ROS_ERROR("Failed to call service get_manipulation_plan");
    return 1;
  }
 
//  manipulation->plan_status = nh.advertise<moveit_task_constructor_msgs::ExecuteTaskSolutionActionResult>("/execute_task_solution/result", 1, true);


  // Keep introspection alive
  ros::waitForShutdown();
  return 0;
}

File: manipulation/scripts/manipulation_request.py
#!/usr/bin/env python3

import rospy
from manipulation.srv import GetManipulationPlan  # Import correct service type
from manipulation.msg import ManipulationPlanRequest
from actionlib.simple_action_client import SimpleActionClient
from moveit_task_constructor_msgs.msg import ExecuteTaskSolutionAction, ExecuteTaskSolutionGoal
from geometry_msgs.msg import Pose, Point, Quaternion  # Import Pose, Point, and Quaternion
from std_msgs.msg import Duration
from moveit_msgs.msg import MoveItErrorCodes

LOGNAME = "manipulation_request"

def main():
    rospy.init_node("manipulation_request")
    # Use a context manager for the spinner
    get_plan = rospy.ServiceProxy("get_manipulation_plan", GetManipulationPlan)  # Fix service type
    manipulation_request = ManipulationPlanRequest()
    manipulation_request.task_type = ManipulationPlanRequest.WIPE
    manipulation_request.object_name = "sponge"
    manipulation_request.support_surfaces = ["table1"]
    manipulation_request.task_name = "task_request"

    # Use dictionary-style initialization for Pose
    manipulation_request.place_pose = Pose(
        position=Point(0.8, 0.11, 0.8),
        orientation=Quaternion(0, 0, 0, 1.0)
    )

    response = get_plan(manipulation_request)

    # Use a more descriptive variable name for the action client
    execute_client = SimpleActionClient("execute_task_solution", ExecuteTaskSolutionAction)
    execute_client.wait_for_server()

    execute_goal = ExecuteTaskSolutionGoal()

    if response:
        rospy.loginfo("Plan received, executing plan")
        execute_goal.solution = response.manipulation_plan_response.solution
        execute_client.send_goal_and_wait(execute_goal)
        execute_result = execute_client.get_result().error_code
        if execute_result.val == MoveItErrorCodes.SUCCESS:
            rospy.loginfo("Execution Success")
            return 0
        else:
            rospy.loginfo("Execution Failed")
            return 1
    else:
        rospy.logerror("Failed to call service get_manipulation_plan")
        return 1



if __name__ == "__main__":
    main()
